### PR TITLE
feat: Allow creation of saml2 integrations (#616)

### PR DESCRIPTION
**Pull request by intinfinite**
- Allows creation of security integrations with type=SAML2 ([example](https://docs.snowflake.com/en/sql-reference/sql/create-security-integration.html#id6))

## Test Plan
Created integration and acceptance tests

## References
This PR should address issue #616 
https://docs.snowflake.com/en/sql-reference/ddl-user-security.html#security-integrations
https://docs.snowflake.com/en/sql-reference/sql/create-security-integration.html#id6

### CODE REVIEW

**Comment by intinfinite** [comment]
> Integration tests failure for [45875c5](https://github.com/chanzuckerberg/terraform-provider-snowflake/actions/runs/1659742853)

Is this an issue on my end?

**Comment by alldoami** [comment]
@intinfinite yes, you can see the logs if you click on the link in the comment.

**Comment by intinfinite** [comment]
Thanks, I think these should fix it

**Comment by intinfinite** [comment]
Looks like it's failing because a SAML2 integration already exists on the account.

```
=== RUN   TestAcc_SamlIntegration
    saml_integration_acceptance_test.go:20: Step 1/1 error: Error running apply: exit status 1
        
        Error: error creating security integration: 390956 (22023): There can only be one enabled SAML2 integration
        
          on terraform_plugin_test.tf line 2, in resource "snowflake_saml_integration" "test_saml_int":
           2: 	resource "snowflake_saml_integration" "test_saml_int" {
        
        
--- FAIL: TestAcc_SamlIntegration (2.40s)
```

Do you know if there is an active SAML2 integration on the runner's Snowflake account the tests are run on or if the integration I created wasn't dropped since the first test failed?

You can check by seeing if any records are returned when running the below with ACCOUNTADMIN:
```sql
show security integrations;
```

If there is a SAML2 integration being used on the account, I can change the test to create a disabled security integration instead.

Thanks!

**Comment by alldoami** [comment]
Hi @intinfinite! Thanks for the thorough comment. Yes we are using the SAML2 Integration, so the test to disable sounds good.

**Comment by intinfinite** [comment]
Sorry for the multiple iterations.  Think this should be the last one *fingers crossed*.

It looks like there is a remnant `SECURITY INTEGRATION` from one of the previous failures that's preventing this from being run and would need to be manually dropped/destroyed before this test will pass.

```
=== RUN   TestAcc_SamlIntegration
    saml_integration_acceptance_test.go:20: Step 1/1 error: Error running apply: exit status 1
        
        Error: error creating security integration: 390957 (22023): Issuer URLs must be distinct across SAML2 integrations.
        
          on terraform_plugin_test.tf line 2, in resource "snowflake_saml_integration" "test_saml_int":
           2: 	resource "snowflake_saml_integration" "test_saml_int" {
        
        
--- FAIL: TestAcc_SamlIntegration (3.09s)
```

Can view if there are any test integrations like this:
```sql
SHOW SECURITY INTEGRATIONS LIKE '%test%';
```

And then drop them like this:
```
DROP SECURITY INTEGRATION "test_saml_integration";
```

**Review by alldoami — CHANGES_REQUESTED** [review]
We cannot test this without modifying our snowflake account's saml integration. There can only be one saml integration at a time. Please do not run this until there are changes made to not create another saml integration.

**Comment by intinfinite** [comment]
You should be able to have multiple security integrations of the same type, just cannot have more than one enabled at a time (see screenshot below).

I updated the test to create a disabled integration, but I think there might be a test integration that's leftover from a previous run that failed before it could be removed.

image

**Comment by alldoami** [comment]
I checked for the `test_saml_integration` but do not see it on my end...

**Comment by intinfinite** [comment]
Thanks for checking.  I think I see what the issue is. 

I used to have the name hard-coded as "test-saml-integration", but changed it to be random to be more in line with the other tests (like the API integration or scim integration).

You can be sure that it's the dummy integration that was created before by describing it using:

```sql
desc integration "jumblednameitgenerated";
```

If the `SAML2_SSO_URL` or `SAML2_ISSUER` has test in it, it's probably the remnant, like below:

image

**Comment by alldoami** [comment]
yup I believe I deleted the one with the random string. Cool that the tests succeeded!

**Comment by intinfinite** [comment]
Awesome, thanks for putting up with me.  First time contributing to this project.

**Comment by alldoami** [comment]
Going to have @edulop91 do a passover, but in the meantime, can you pull in the most recent changes @intinfinite?

**Comment by intinfinite** [comment]
Sounds good. Merged from main.

**Comment by alldoami** [comment]
Hm still looks out of sync with `main`

**Comment by intinfinite** [comment]
All set

File: pkg/resources/saml_integration.go [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+			return errors.Wrap(err, "error updating security integration")
+		}
+	}
+
+	return ReadSAMLIntegration(d, meta)
+}
+
+// DeleteSAMLIntegration implements schema.DeleteFunc
+func DeleteSAMLIntegration(d *schema.ResourceData, meta interface{}) error {
+	return DeleteResource("", snowflake.ScimIntegration)(d, meta)
```
Comment (edulop91) [inline pkg/resources/saml_integration.go:475]: `snowflake.ScimIntegration` seems off -- is this snowflake.SAMLIntegration?

Comment (intinfinite) [reply pkg/resources/saml_integration.go]: good catch, fixed in latest commit

**Review by alldoami — APPROVED** [review]

File: pkg/resources/saml_integration.go (hunk 1 above)
Comment (edulop91) [reply pkg/resources/saml_integration.go]: thank you!

**Pull request merged**